server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -88,11 +88,13 @@ public IndexerSQLMetadataStorageCoordinator(
             final ResultIterator<byte[]> dbSegments =
                 handle.createQuery(
                     String.format(
-                        "SELECT payload FROM %s WHERE used = true AND dataSource = :dataSource",
+                        "SELECT payload FROM %s WHERE used = true AND dataSource = :dataSource AND start >= :start and \"end\" <= :end  AND used = true",
```

Comment:
it now matches behavior of `getUnusedSegmentsForInterval`. The prior behavior returned ALL segments REGARDLESS of start and end, which is why I made the change here. 

After thinking a bit more, the correct solution probably is closer to what you are suggesting, where a segment returns if it overlaps with the segment start/end timeframe.

I'm assuming `getUnusedSegmentsForInterval` should also follow this behavior.